### ISSUE TITLE
Integration tests got stuck

### ISSUE THREAD

**Issue by calohmn** [eclipse-hono/hono#3321]
Today there was an integration test run that got stuck.

Last log output:
```
2022-06-28T09:58:49.2283580Z 09:58:49.222 [main] INFO  o.e.h.tests.mqtt.TelemetryMqttQoS0IT - running testUploadMessagesViaGateway(VertxTestContext)
2022-06-28T09:58:49.2361887Z 09:58:49.228 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:36951
2022-06-28T09:58:49.2369980Z 09:58:49.236 [main] INFO  o.e.h.tests.IntegrationTestSupport - Configured to connect to Kafka on localhost:36951
2022-06-28T09:58:49.2628397Z 09:58:49.258 [main] INFO  o.e.h.s.a.SpringBasedHonoPasswordEncoder - using BCrypt [strength: 4] with PRNG [NativePRNGNonBlocking] for encoding clear text passwords
2022-06-28T09:58:49.4836174Z 09:58:49.483 [vert.x-eventloop-thread-1] INFO  o.e.h.s.a.SpringBasedHonoPasswordEncoder - using BCrypt [strength: 4] with PRNG [NativePRNGNonBlocking] for encoding clear text passwords
2022-06-28T09:58:49.9056973Z 09:58:49.903 [vert.x-eventloop-thread-1] INFO  o.e.h.tests.mqtt.TelemetryMqttQoS0IT - MQTTS connection to adapter [host: localhost, port: 49173] established
2022-06-28T09:58:50.3824290Z 09:58:50.381 [vert.x-eventloop-thread-1] INFO  o.e.h.tests.mqtt.TelemetryMqttQoS0IT - messages sent: 50
2022-06-28T09:58:50.6737120Z 09:58:50.673 [vert.x-eventloop-thread-1] INFO  o.e.h.tests.mqtt.TelemetryMqttQoS0IT - messages sent: 100
2022-06-28T09:58:50.6974266Z 09:58:50.696 [vert.x-eventloop-thread-1] ERROR o.e.h.tests.mqtt.TelemetryMqttQoS0IT - error sending message 104
2022-06-28T09:58:50.6975424Z io.netty.handler.ssl.SslClosedEngineException: SSLEngine closed already
2022-06-28T09:58:50.6976135Z 	at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:859)
2022-06-28T09:58:50.6976825Z 	at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:798)
2022-06-28T09:58:50.6977636Z 	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:779)
2022-06-28T09:58:50.6978358Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
2022-06-28T09:58:50.6979184Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:742)
2022-06-28T09:58:50.6979948Z 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:728)
2022-06-28T09:58:50.6980849Z 	at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:313)
2022-06-28T09:58:50.6981691Z 	at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:131)
2022-06-28T09:58:50.6982850Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
2022-06-28T09:58:50.7098719Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:742)
2022-06-28T09:58:50.7116983Z 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:728)
2022-06-28T09:58:50.7117662Z 	at io.netty.channel.ChannelDuplexHandler.flush(ChannelDuplexHandler.java:127)
2022-06-28T09:58:50.7118320Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
2022-06-28T09:58:50.7119089Z 	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:765)
2022-06-28T09:58:50.7119878Z 	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:790)
2022-06-28T09:58:50.7121506Z 	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:758)
2022-06-28T09:58:50.7122311Z 	at io.vertx.core.net.impl.ConnectionBase.write(ConnectionBase.java:177)
2022-06-28T09:58:50.7122867Z 	at io.vertx.core.net.impl.ConnectionBase.writeToChannel(ConnectionBase.java:229)
2022-06-28T09:58:50.7123450Z 	at io.vertx.core.net.impl.ConnectionBase.writeToChannel(ConnectionBase.java:214)
2022-06-28T09:58:50.7124021Z 	at io.vertx.core.net.impl.ConnectionBase.writeToChannel(ConnectionBase.java:210)
2022-06-28T09:58:50.7124593Z 	at io.vertx.core.net.impl.NetSocketImpl.writeMessage(NetSocketImpl.java:136)
2022-06-28T09:58:50.7125331Z 	at io.vertx.core.net.impl.NetSocketImpl.writeMessage(NetSocketImpl.java:130)
2022-06-28T09:58:50.7125953Z 	at io.vertx.mqtt.impl.MqttClientImpl.write(MqttClientImpl.java:863)
2022-06-28T09:58:50.7126482Z 	at io.vertx.mqtt.impl.MqttClientImpl.publish(MqttClientImpl.java:445)
2022-06-28T09:58:50.7127013Z 	at io.vertx.mqtt.impl.MqttClientImpl.publish(MqttClientImpl.java:454)
2022-06-28T09:58:50.7127755Z 	at org.eclipse.hono.tests.mqtt.TelemetryMqttQoS0IT.lambda$send$0(TelemetryMqttQoS0IT.java:65)
2022-06-28T09:58:50.7128342Z 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:893)
2022-06-28T09:58:50.7128863Z 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:860)
2022-06-28T09:58:50.7129380Z 	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
2022-06-28T09:58:50.7129845Z 	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:274)
2022-06-28T09:58:50.7130326Z 	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
2022-06-28T09:58:50.7130828Z 	at io.vertx.core.impl.AbstractContext.emit(AbstractContext.java:53)
2022-06-28T09:58:50.7131332Z 	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
2022-06-28T09:58:50.7131845Z 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.run(VertxImpl.java:883)
2022-06-28T09:58:50.7132365Z 	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
2022-06-28T09:58:50.7132904Z 	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
2022-06-28T09:58:50.7133574Z 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
2022-06-28T09:58:50.7134273Z 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
2022-06-28T09:58:50.7134869Z 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
2022-06-28T09:58:50.7135429Z 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
2022-06-28T09:58:50.7136015Z 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
2022-06-28T09:58:50.7136575Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2022-06-28T09:58:50.7137079Z 	at java.base/java.lang.Thread.run(Thread.java:833)
2022-06-28T09:58:50.7137858Z 09:58:50.705 [vert.x-eventloop-thread-1] ERROR io.vertx.core.impl.ContextImpl - Unhandled exception
2022-06-28T09:58:50.7138660Z java.lang.NullPointerException: Cannot invoke "io.vertx.core.net.impl.NetSocketInternal.writeMessage(Object)" because the return value of "io.vertx.mqtt.impl.MqttClientImpl.connection()" is null
2022-06-28T09:58:50.7139390Z 	at io.vertx.mqtt.impl.MqttClientImpl.write(MqttClientImpl.java:863)
2022-06-28T09:58:50.7139902Z 	at io.vertx.mqtt.impl.MqttClientImpl.publish(MqttClientImpl.java:445)
2022-06-28T09:58:50.7140418Z 	at io.vertx.mqtt.impl.MqttClientImpl.publish(MqttClientImpl.java:454)
2022-06-28T09:58:50.7140970Z 	at org.eclipse.hono.tests.mqtt.TelemetryMqttQoS0IT.lambda$send$0(TelemetryMqttQoS0IT.java:65)
2022-06-28T09:58:50.7141552Z 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:893)
2022-06-28T09:58:50.7142073Z 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:860)
2022-06-28T09:58:50.7142588Z 	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
2022-06-28T09:58:50.7143287Z 	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:274)
2022-06-28T09:58:50.7143769Z 	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
2022-06-28T09:58:50.7144246Z 	at io.vertx.core.impl.AbstractContext.emit(AbstractContext.java:53)
2022-06-28T09:58:50.7144752Z 	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
2022-06-28T09:58:50.7145261Z 	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.run(VertxImpl.java:883)
2022-06-28T09:58:50.7145776Z 	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
2022-06-28T09:58:50.7146331Z 	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
2022-06-28T09:58:50.7146964Z 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
2022-06-28T09:58:50.7147842Z 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
2022-06-28T09:58:50.7148397Z 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
2022-06-28T09:58:50.7158050Z 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
2022-06-28T09:58:50.7161763Z 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
2022-06-28T09:58:50.7162449Z 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2022-06-28T09:58:50.7168974Z 	at java.base/java.lang.Thread.run(Thread.java:833)
2022-06-28T09:58:52.2851066Z 09:58:52.284 [vert.x-eventloop-thread-1] INFO  o.e.h.tests.mqtt.TelemetryMqttQoS0IT - messages received: 50
2022-06-28T09:58:53.1692232Z 09:58:53.168 [vert.x-eventloop-thread-1] INFO  o.e.h.tests.mqtt.TelemetryMqttQoS0IT - messages received: 100
2022-06-28T10:57:46.4417681Z ##[error]The operation was canceled.
```

The unhandled exception on `MqttClientImpl.publish` lead to a future not getting completed.

**Comment by avgustinmm** [comment]
in MqttPublishTestBase there are few things that could be improved/fixed:
- messageHandlingCompleted.await(); should wait with certain timeout (as received.await does)
- in doTestUploadMessages if sendAttempt.failed() is true and errorMessagesExpected is false (as in this case) the the fail is just logged ("error sending message {}") and nothing is done. In this case the test shall fail
- the NPE on the following (?) send attempt seems more like a bug in MqttClientImpl. It should check that the client is null and return failed feature instead of throwing NPE. It may be reported as a bug in the Vertx MQTT project but nevertheless that exception could be handled in the test (by try / catch) which combined with previous should prevent even such problems

**Comment by sophokles73** [comment]
@avgustinmm sounds reasonable, would you mind creating corresponding PR(s)?

**Comment by avgustinmm** [comment]
OK, I'll try to create a PR

**Comment by sophokles73** [comment]
closed via #3472
